cmds/tail/tail.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Tail prints the lasts 10 lines of a file. Can additionally follow the
+// the end of the file as it grows.
+//
+// Synopsis:
+//     tail [-f] [-n lines_to_show] [FILE]
+//
+// Description:
+//     If no files are specified, read from stdin.
+//
+// Options:
+//     -f: follow the end of the file as it grows
+//     -n: specify the number of lines to show (default: 10)
+
+// Missing features:
+// - follow-mode (i.e. tail -f)
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"syscall"
+)
+
+var (
+	flagFollow   = flag.Bool("f", false, "follow the end of the file")
+	flagNumLines = flag.Int("n", 10, "specify the number of lines to show")
+)
+
+// TailConfig is a configuration object for the Tail function
+type TailConfig struct {
+	follow   bool // enable follow-mode (-f)
+	numLines uint // specifies the number of lines to print (-n)
+}
+
+// getBlockSize returns the number of bytes to read for each ReadAt call. This
+// helps minimize the number of syscalls to get the last N lines of the file.
+func getBlockSize(numLines uint) int64 {
+	// This is currently computed as 81 * N, where N is the requested number of
+	// lines, and 81 is a relatively generous estimation of the average line
+	// length.
+	return 81 * int64(numLines)
+}
+
+// readLastLinesBackwards reads the last N lines from the provided file, reading
+// backwards from the end of the file. This is more efficient than reading from
+// the beginning, but can only be done on seekable files, (e.g. this won't work
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile *os.File, writer io.Writer, numLines uint) error {
+	blkSize := getBlockSize(numLines)
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
+	if err != nil {
+		return err
+	}
+	// read block by block backwards until `numLines` lines are found
+	readData := make([]byte, 0)
+	buf := make([]byte, blkSize)
+	pos := lastPos
+	var foundLines uint
+	// for each block, count how many new lines, until they add up to `numLines`
+	for {
+		if pos == 0 {
+			break
+		}
+		var thisChunkSize int64
+		if pos < blkSize {
+			thisChunkSize = pos
+		} else {
+			thisChunkSize = blkSize
+		}
+		pos -= thisChunkSize
+		n, err := inFile.ReadAt(buf, pos)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		// merge this block to what was read so far
+		readData = append(buf[:n], readData...)
+		// count how many lines we have so far, and stop reading if we have
+		// enough
+		foundLines += uint(bytes.Count(buf[:n], []byte{'\n'}))
+		if foundLines >= numLines {
+			break
+		}
+	}
+	// find the start of the n-th to last line
```

Comment:
Maybe this could be in its own function? There are a lot of slices going around here.

```go
func lastNLines(buf []byte, n int) []byte {
  
}
```

Reply:
fixed in the next commit